packages/wallets/solflare/src/metamask/detect.ts

```diff
@@ -0,0 +1,59 @@
+import type { EthereumProvider, WindowWithEthereum } from '@solflare-wallet/metamask-sdk';
+import { registerWallet } from '@wallet-standard/wallet';
+import { SolflareMetaMaskWallet } from './wallet.js';
+
+let stopPolling = false;
+
+/** @internal */
+export function detectAndRegisterSolflareMetaMaskWallet(): boolean {
+    // If detected, stop polling.
+    if (stopPolling) return true;
+    (async function () {
+        try {
+            // Try to detect, stop polling if detected, and register the wallet.
+            if (await isSnapProviderDetected()) {
+                stopPolling = true;
+                registerWallet(new SolflareMetaMaskWallet());
+            }
+        } catch (error) {
+            // Stop polling on unhandled errors (this should never happen).
+            stopPolling = true;
+        }
+    })();
+    // Keep polling.
+    return false;
+}
+
+async function isSnapProviderDetected(): Promise<boolean> {
+    try {
+        const provider = (window as WindowWithEthereum).ethereum;
+        if (!provider) return false;
+
+        const providerProviders = provider.providers;
+        if (providerProviders && Array.isArray(providerProviders)) {
+            for (const provider of providerProviders) {
+                if (await isSnapSupported(provider)) return true;
+            }
+        }
+
+        const providerDetected = provider.detected;
+        if (providerDetected && Array.isArray(providerDetected)) {
+            for (const provider of providerDetected) {
+                if (await isSnapSupported(provider)) return true;
+            }
+        }
+
+        return await isSnapSupported(provider);
```

Comment:
Is there any way to make this detection more efficient? How are `provider.providers` and `provider.detected` used such that we have to scan them all?

The only problem I have with this approach now is that we're polling and we never stop polling unless it's detected. At some point, maybe we _should_ stop, if the provider/snap support isn't detected after some time/attempts.

Reply:
All EVM wallets inject their API to `window.ethereum`. Some wallets keep the provider that was injected before them to fields like `window.ethereum.providers` (Coinbase) or `window.ethereum.detected` (Phantom), while others just replace the previous one with their own interface.

Reply:
Got it, thanks for clarifying!